test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,294 @@
+
+	var (
+		ctx = context.Background()
+		cfg = dockerchain.DefaultConfig(s.client, s.network).WithTag(ImageTag)
+		kr  = cfg.Genesis.Keyring()
+	)
+	cfg.Genesis = cfg.Genesis.WithAppVersion(TargetAppVersion - 1)
+
+	chain, err := dockerchain.NewCelestiaChainBuilder(s.T(), cfg).Build(ctx)
+	s.Require().NoError(err)
+
+	s.T().Cleanup(func() {
+		if err := chain.Stop(ctx); err != nil {
+			s.T().Logf("Error stopping chain: %v", err)
+		}
+	})
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
+
+	// Sanity check: Test bank send before upgrade
+	s.T().Log("Testing bank send functionality before upgrade")
+	testBankSend(s.T(), chain, cfg)
+
+	validatorNode := chain.GetNodes()[0]
+	rpcClient, err := validatorNode.GetRPCClient()
+	s.Require().NoError(err, "failed to get RPC client")
+
+	abciInfo, err := rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+	currentAppVer := abciInfo.Response.GetAppVersion()
+	s.T().Logf("Current app version before upgrade: %d", currentAppVer)
+
+	records, err := kr.List()
+	s.Require().NoError(err, "failed to list accounts")
+	s.Require().Len(records, len(chain.GetNodes()), "number of accounts does not match number of nodes")
+
+	// Signal for upgrade and get the upgrade height
+	upgradeHeight := s.signalAndGetUpgradeHeight(ctx, chain, validatorNode, records, TargetAppVersion)
+
+	// Get current height
+	status, err := rpcClient.Status(ctx)
+	s.Require().NoError(err, "failed to get node status")
+	currentHeight := status.SyncInfo.LatestBlockHeight
+
+	s.T().Logf("Current height: %d, Upgrade height: %d", currentHeight, upgradeHeight)
+
+	// Wait until we reach the upgrade height
+	blocksToWait := int(upgradeHeight-currentHeight) + 2 // Add buffer
+	s.T().Logf("Waiting for %d blocks to reach upgrade height plus buffer", blocksToWait)
+	s.Require().NoError(wait.ForBlocks(ctx, blocksToWait, chain))
+
+	// Verify the app version has been upgraded
+	abciInfo, err = rpcClient.ABCIInfo(ctx)
+	s.Require().NoError(err, "failed to fetch ABCI info")
+
+	// Verify app version is upgraded
+	s.Require().Equal(TargetAppVersion, abciInfo.Response.GetAppVersion(), "app_version mismatch")
+
+	// Sanity check: Test bank send after upgrade
+	s.T().Log("Testing bank send functionality after upgrade")
+	testBankSend(s.T(), chain, cfg)
+}
+
+// signalAndGetUpgradeHeight signals for an upgrade to the specified app
+// version and returns the scheduled upgrade height.
+func (s *CelestiaTestSuite) signalAndGetUpgradeHeight(ctx context.Context, chain tastoratypes.Chain, validatorNode tastoratypes.ChainNode, records []*keyring.Record, targetAppVersion uint64) int64 {
+	// Signal for the upgrade using builder
+	for i, node := range chain.GetNodes() {
+		s.T().Logf("Signaling for upgrade to version %d from validator %d", targetAppVersion, i)
+
+		signalCmd := []string{"tx", "signal", "signal", fmt.Sprintf("%d", targetAppVersion), "--from", records[i].Name}
```

Comment:
how do we know that the keyring is in the same order as chain.GetNodes?

if that ever changes, presumably this will break? maybe we should just change this in a flup to use txclient to avoid this issue.